crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -584,39 +592,49 @@ impl<TX: DbTxMut + DbTx> DatabaseProvider<TX> {
         let mut plain_accounts_cursor = self.tx.cursor_write::<tables::PlainAccountState>()?;
         let mut plain_storage_cursor = self.tx.cursor_dup_write::<tables::PlainStorageState>()?;
 
-        let mut reverts: RevertsInit = HashMap::new();
-
         // add account changeset changes
         for (block_number, account_before) in account_changeset.into_iter().rev() {
             let AccountBeforeTx { info: old_info, address } = account_before;
-            match state.entry(address) {
-                hash_map::Entry::Vacant(entry) => {
-                    let new_info = plain_accounts_cursor.seek_exact(address)?.map(|kv| kv.1);
-                    entry.insert((old_info, new_info, HashMap::new()));
-                }
-                hash_map::Entry::Occupied(mut entry) => {
-                    // overwrite old account state.
-                    entry.get_mut().0 = old_info;
+            if !bundle_builder.get_states().contains(&address) {
+                let new_info = plain_accounts_cursor.seek_exact(address)?.map(|kv| kv.1);
+                if let Some(new_info) = new_info {
+                    bundle_builder =
+                        bundle_builder.state_present_account_info(address, into_revm_acc(new_info));
                 }
             }
+            if let Some(old_info) = old_info {
+                bundle_builder =
+                    bundle_builder.state_original_account_info(address, into_revm_acc(old_info));
+            }
```

Comment:
previous behavior: if the entry is occuppied, overwrite the original account state even if `old_info` is `None`
current behavior: only overwrite the original account state if `old_info` is `Some`

if `old_info` is `None`, original info must still be overwritten